Fix for issue #8 - Out-of-bounds read of VIN buffer

marcr opened the pull request:
See issue #8 for bug report.

The fix checks the VIN length and adds padding of '0' characters in case the VIN value is shorter than 17 bytes.
Added unit tests for the padding code as well.

Additionally enabled ASan in the Makefile and fixed two further memleaks in other tests.

marcr committed:
Issue #8 - Add missing length check when reading VIN

If VIN is shorter than 16 bytes, the message gets padded with zero bytes.

marcr committed:
Issue 8 - Add test cases for VIN padding

marcr committed:
Fix memleaks in unit tests.

Both tests were missing a TearDown() to free the memory reserved in
SetUp().

tjaehnel commented on libdoipserver/src/VehicleIdentificationHandler.cpp:
```diff
@@ -1,20 +1,23 @@
+            j++;
+	} else {
+	    //Pad with zero if VIN is shorter than 16 bytes
+	    message[i] = 0;
```
Not sure if it makes sense to pad with 0.
DoIP does not define the VIN as a Zero-Terminated string, so a client will likely use all 17 (not 16 as written in the comment!) bytes as chars.
I would suggest to either terminate the whole process on startup of the server when the length of the configured VIN is not 17.
Or pad with one of the characters allowed for a VIN (German wikipedia has a nice list). I suggest 'X' or '0'.

marcr replied:
I figured, if the VIN is ASCII, a padding with \0 would make it easier to parse it as string. But other symbols would be fine for me, too.
Oh, and thanks for pointing out the 17 bytes.

If we terminate the process on startup in case of an invalid VIN format, then the example code needs to be fixed, too. As that was referenced in the bug report.

tjaehnel replied:
This code prepares the VIN for transmitting. ISO13400-2 defines a fixed 17 bytes length for the VIN and no zero termination. Therefore we should not zero-terminate the transmitted VIN but always send 17 valid characters.

marcr replied:
Updated the code to use '0' instead. And fixed the length naming.
Also added a comment to the example code.

marcr committed:
Issue #8 - Change VIN Padding to valid ASCII char

- Fix comments and names referring to 16 VIN bytes. Use 17 instead.
- Add comment to example code mentioning the padding.
